vermouth/processors/apply_rubber_band.py

```diff
@@ -0,0 +1,120 @@
+    return np.sqrt(((coordinates[:, np.newaxis, :] - coordinates[np.newaxis, :, :]) ** 2).sum(axis=-1))
+
+
+def compute_decay(distance_matrix, shift, rate, power):
+    return np.exp(-rate * ((distance_matrix - shift) **  power))
+
+
+def compute_force_constants(distance_matrix, lower_bound, upper_bound,
+                            decay_factor, decay_power, base_constant,
+                            minimum_force):
+    constants = compute_decay(distance_matrix, lower_bound, decay_factor, decay_power)
+    np.fill_diagonal(constants, 0)
+    constants *= base_constant
+    constants[constants < minimum_force] = 0
+    constants[distance_matrix > upper_bound] = 0
+    return constants
+
+
+def build_connectivity_matrix(graph, separation, selection=None):
+    if separation <= 0:
+        raise ValueError('Separation has to be strictly positive.')
+    if separation == 1:
+        return nx.to_numpy_matrix(graph, nodelist=selection).astype(bool)
+    subgraph = graph.subgraph(selection)
+    connectivity = np.zeros((len(subgraph), len(subgraph)), dtype=bool)
+    for (i, key_i),  (j, key_j) in itertools.combinations(enumerate(subgraph.nodes), 2):
+        shortest_path = len(nx.shortest_path(subgraph, key_i, key_j))
+        # The source and the target are counted in the shortest path
+        connectivity[i, j] = shortest_path <= separation + 2
+        connectivity[j, i] = connectivity[i, j]
+    return connectivity
+
+
+
+def apply_rubber_band(molecule, selector,
+                      lower_bound, upper_bound,
+                      decay_factor, decay_power,
+                      base_constant, minimum_force,
+                      bond_type):
+    selection = []
+    coordinates = []
+    missing = []
+    for node_key, attributes in molecule.nodes.items():
+        if selector(attributes):
+            selection.append(node_key)
+            coordinates.append(attributes.get('position'))
+            if coordinates[-1] is None:
+                missing.append(node_key)
+    if missing:
+        raise ValueError('All atoms from the selection must have coordinates. '
+                         'The following atoms do not have some: {}.'
+                         .format(' '.join(missing)))
+    coordinates = np.stack(coordinates)
+    distance_matrix = self_distance_matrix(coordinates)
+    constants = compute_force_constants(distance_matrix, lower_bound,
+                                        upper_bound, decay_factor, decay_power,
+                                        base_constant, minimum_force)
+    connectivity = build_connectivity_matrix(molecule, 2, selection=selection)
+    constants *= ~connectivity
+    for from_idx, to_idx in zip(*np.triu_indices_from(constants)):
+        from_key = selection[from_idx]
+        to_key = selection[to_idx]
+        force_constant = constants[from_idx, to_idx]
+        length = round(distance_matrix[from_idx, to_idx], 5)
+        if force_constant > minimum_force:
```

Comment:
Should this check not be part of `compute_force_constants`?

Reply:
`compute_force_constants` compute the constants, what we do with them is not its responsibility. Also, filtering the force constants in `compute_force_constants` will only change the condition of the test, it will not get rid of it.

Reply:
In `make_force_constants` there's already a check that does
`constants[constants < minimum_force] = 0`
So maybe just `if force_constant`?

Reply:
Indeed. It is possible, but then the way it is here may be more explicit.

Reply:
Hmmn. Or set it to np.nan in compute_force_constants, and check for that here. Now you're doing the same thing twice. And set a comment that it's set to nan in said function